beps/0004-scaffolder-task-idempotency/README.md

```diff
@@ -259,6 +260,26 @@ Task state will be stored in the extra column `state` in the table `tasks` with
 }
 ```
 
+### Rolling back to an initial state
+
+We will not provide built in logic which will do that magically for you, instead we will give an opportunity to hook own
+logic where user can do a cleanup of created resources. Each action definition will have an additional optional property
+`undoAction`, which will be triggered during 'Rollback'.
+
+```yaml
+- id: publish
+  name: Publish
+  action: publish:github
+  undoAction: delete:github
```

Comment:
I think we should think about the different options here. Whilst this keeps the actions pretty clean, it could be that actually it doesn't bloat them too much.

Was thinking over the weekend about this `publish:github` action in particular, there's only really one thing to undo if we need to which is basically to just delete the repository totally. We don't need to undo the `collaborators` and all the smaller things individually when we can just delete the repository.

Reply:
Also wonder if we should consider this a separate BEP entirely, as to not bloat the existing proposal scope?

Reply:
Yeah indeed, that's why we don't need to have a revert for each checkpoint, but rather 1 for entire action.

I was thinking to make it work with a little effort possible, so we can fit it in.

Like for example, when user runs 'Roll back', and instead of `action` we do run `undoAction`

Reply:
Thinking that `rollback` is optional. And that you will maybe have a different `ctx` object in rollback. The reason being you probably want access to the current checkpoints state so that you could potentially do different things depending on how far you got through the `handler`.

The idea would be that the `rollback` handler would just be something like:

```
check if repo was created by the scaffolder job -> delete repo -> done
```

Reply:
Some of this logic of course can be broken out into shared functions, but thinking it makes sense to keep the implementation close to the actions in the first place, rather than them being totally separate actions. Then you have `undoActions` which are closely coupled to the original `action` in the first place. It's a little unfortunate to hide that connection I'm thinking.

Reply:
Okie, how about then to cover 'publish:github' in this BEP to validate the approach and to cover the rest of actions by community contributions?

Reply:
I'm still feeling that we should probably create another BEP for this instead of adjusting the scope for the original BEP. Feels like they're two related, but fundamentally different features.